File: T-Bank/Algorithms-and-Data-Structures-2025/6sprint-Graph1/contest/C-Топологическая_сортировка/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(|V| + |E|)
    Memory Complexity O(|V| + |E|)
    """
    n, m = map(int, sys.stdin.readline().rstrip().split())
    adj_list: list[list[int]] = [[] for _ in range(n + 1)]
    for _ in range(m):
        u, v = map(int, sys.stdin.readline().rstrip().split())
        adj_list[u].append(v)
    order: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    positions: list[int] = [0] * (n + 1)
    for i in range(n): positions[order[i]] = i
    for source in range(1, n + 1):
        for destination in adj_list[source]:
            if positions[source] > positions[destination]:
                sys.stdout.write('NO\n')
                return
    sys.stdout.write('YES\n')


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/6sprint-Graph1/contest/C-Топологическая_сортировка/solution.cpp
#include <iostream>
#include <vector>
#include <stack>

void solution() {
    /*
    Time Complexity O(|V| + |E|)
    Memory Complexity O(|V| + |E|)
    */
    int n, m;
    std::scanf("%d %d", &n, &m);
    std::vector<std::vector<int>> adjList(n + 1);
    for (int i = 0; i < m; ++i) {
        int u, v;
        std::scanf("%d %d", &u, &v);
        adjList[u].push_back(v);
    }
    std::vector<int> order(n);
    for (int i = 0; i < n; ++i) std::scanf("%d", &order[i]);
    std::vector<int> position(n + 1);
    for (int i = 0; i < n; ++i) position[order[i]] = i;
    for (int u = 1; u <= n; ++u) {
        for (int v : adjList[u]) {
            if (position[u] > position[v]) {
                std::printf("NO\n");
                return;
            }
        }
    }

    std::printf("YES\n");
}

int main() {
    solution();
    return 0;
}